Update c2.vim

Angluca opened the pull request:
Update c2.vim
- add \<c-h\> use c2tags goto func header 
- add plugin\/indent fix some \<cr\> goto line pos issue

Angluca committed:
Update syntax

Angluca committed:
Fix error for Plug.vim

Angluca committed:
Add tags and dicts, you can autocomple typevar in vim

Angluca commented:
```
https://github.com/Angluca/AcVim/blob/master/vimfiles/dict/c2.ctags
ctags --options=./c2.ctags  -R -f c2.tags
```
This is ctags.conf, You can use it generat tags and must insert PATH to tags, now you can ```<c-]>``` goto the file, ```<c-x><c-k>```can complete word when you can't autocomplete

bvdberg commented:
C2 has a compiler-generated refererences file that guarantees a 100% correct lookup of a definition and uses of a certain symbol. To use that, a simple plugin is needed to interface between the editor and the c2tags binary. It works a LOT faster than ctags and only shows correct matches

Angluca commented:
Hi, How to use c2tags complete in vim, I don't know Tell me please T_T @bvdberg

Angluca committed:
Remve tags

bvdberg commented:
I'll also add this somewhere in a Readme.
- clone git@github.com:c2lang/c2.vim.git  For me i put it in ~/.vim/pack/plugins/start/
  my plugin manager then auto starts it
- add this to .vimrc:
```
nnoremap <silent> <c-h> :call GetTagResult()<CR>

function! GetTagResult()
  " get cursor position and word
  let l:src_file = bufname('%')
  let l:line = line('.')
  let l:column = col('.')

  " search reference destination
  let l:cmd = 'c2tags ' . l:src_file . ' ' . l:line . ' ' . l:column
  let l:result = substitute(system(l:cmd), '[\]\|[[:cntrl:]]', '', 'g')

  " parse results
  let l:words = split(l:result, ' ')
  if (l:words[0] == 'error')
      echo l:result
      return
  endif
  if (l:words[0] == 'multiple')
      echo "multiple matches found. not supported yet"
      return
  endif
  if (len(l:words) != 4)
      "echo "no results found"
      return
  endif

  " otherwise c2tags returns: 'found file line col'
  let l:dst_file = l:words[1]
  let l:dst_line = l:words[2]
  let l:dst_col = l:words[3]

  " add position to jumplist
  normal! m`

  let l:nr = bufnr("")
  if (l:src_file != l:dst_file)
      let l:nr = bufnr(l:dst_file, 1)
      "let l:nr = bufnr("bar.c2", 1)
      if (l:nr == -1)
          echo "error creating buffer"
          return
      endif
      execute ":buffer ".l:nr
  endif

  " TODO push to tag stack for Ctrl-T
  let l:res = setpos('.', [0, l:dst_line, l:dst_col, 0])
  if (l:res == -1)
      echo "c2tags: cannot set position to:" l:dst_file l:dst_line l:dst_col
      return
  endif
endfunction
```

- compile c2c and make sure c2tags is in your PATH
- compile your project with c2c (enabling the refs plugin) in the recipe:
`plugin deps_generator [all-targets files private]`

Now pressing ctrl-h will jump to a symbol, ctrl-o to go back (in a stack like manner)

So the vim plugin only does the following:
- call c2tags <file> <file> <col>
- parse the reply and jump to it

You can see how c2tags works by calling it manually also

As a sidenote, c2rename uses the same dictionary and can do global renames of symbols

Angluca commented:
Nice!! It can bindkey <c-]>  goto header like ctags
c2rename how to use? I don't know what's this T_T

bvdberg commented:
The Vim glue code was just my first attempt a linking Vim with c2tags. I also have this working for neovim using a lua plugin. On the C2 roadmap is a LSP (Language Server) that would take care of this all in a less haphazard way ;)

Angluca commented:
I with ai use c code write c2_lsp_server, Maybe can use c2 write it like clangd, But I don't know about the JSON protocol for lsp
[c2_lsp_server_c.zip](https://github.com/user-attachments/files/22658197/c2_lsp_server_c.zip)

I'm a c2 newbie, Hope use c2 write game and Can C2 compile wasm in the future?
At present, I can use the C2 build generate C compile wasm with emcc, but it is too troublesome must comment more c_type lineT_T, I hope that C2 will support compiling wasm in the future

bvdberg commented:
I've written some graphics using SDL2 in C2, some tile-based stuff. That works fine. Also networking is supporting everything needed, even supports io_uring on linux for handling many many events per second. I'm currently writing my own backend, targetting arm64, amd4 and rv64. Wasm might be later, but the first 3 will keep me occupied for some time :)

Angluca commented:
Okay, I will learn C2 language to write things first, the grammar and ideas of C2 are really great and concise, I like it very much, I hope C2 will be better in the future

-----
Why not test C3? Because c3 requires clang16 to be used, old machines can't be used at all, and More people can't even try it, so I think c3 is not suitable for the public, and It don't even give the old coder a chance to try haha

Angluca committed:
Add <c-h> use c2tags ..

Angluca commented:
HI, I want some C libraries header change to C2,
What tools or methods can C2 quickly convert C header files to C2i header files?
I don't seem to see this in the tutorial, if you manually change the C header file to C2, it should be very time-consuming

Angluca committed:
Add indent
fixed some <cr> line issues
enum i32 {
    xxx,|<cr>
       >|
}

switch(..) {
    default:|<cr>
   >|
    break;
}

struct {
    i32 : 8;|<cr>
          >|
}

const char[] Multi = "The"|<cr> or { "The"|<cr>
>|

Angluca committed:
update
